File: app/src/main/java/com/example/chap17lifecycleserviceexample/JobIntentService.java
package com.example.chap17lifecycleserviceexample;

import android.app.IntentService;
import android.content.Intent;
import android.util.Log;

import androidx.annotation.Nullable;

public class JobIntentService extends IntentService {
    /**
     * Creates an IntentService.  Invoked by your subclass's constructor.
     *
     * @param name Used to name the worker thread, important only for debugging.
     */
    public JobIntentService() {
        super("name");
    }

    @Override
    public void onCreate() {
        super.onCreate();
        Log.d("Service", "JobIntentService onCreate");
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        Log.d("Service", "JobIntentService onStartCommand");
        Log.d("Service Thread", Thread.currentThread().getName());
        return super.onStartCommand(intent, flags, startId);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        Log.d("Service", "JobIntentService onDestroy");
    }

    @Override
    protected void onHandleIntent(@Nullable Intent intent) {
        Log.d("Service", "JobIntentService onHandleIntent");
    }
}

File: app/src/main/java/com/example/chap7lifecycle/JobIntentService.kt
package com.example.chap7lifecycle

import android.content.Context
import android.content.Intent
import android.util.Log
import androidx.core.app.JobIntentService

class JobIntentService: JobIntentService() {
    override fun onHandleWork(intent: Intent) {
        Log.d("Service", "JobIntentService onHandleWork")
        Log.d("Service Thread", Thread.currentThread().name)
    }

    // We use companion object so that the function can be accessed from external
    companion object {
        fun myBackgroundService(context: Context, intent: Intent) {
            enqueueWork(context, JobIntentService::class.java, 1, intent)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        Log.d("Service", "JobIntentService onDestroy")
    }
}